src/QsCompiler/Transformations/MonomorphizationValidation.cs

```diff
@@ -0,0 +1,69 @@
+﻿using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.MonomorphizationValidation
+{
+    public class MonomorphizationValidationTransformation
+    {
+        public static void Apply(QsCompilation compilation)
+        {
+            var filter = new MonomorphizationValidationSyntax();
+
+            foreach (var ns in compilation.Namespaces)
+            {
+                filter.Transform(ns);
+            }
+        }
+
+        internal class MonomorphizationValidationSyntax : SyntaxTreeTransformation<ScopeTransformation<MonomorphizationValidationExpression>>
+        {
+            public MonomorphizationValidationSyntax(ScopeTransformation<MonomorphizationValidationExpression> scope = null) :
+                base(scope ?? new ScopeTransformation<MonomorphizationValidationExpression>(new MonomorphizationValidationExpression())) { }
+
+            public override ResolvedSignature onSignature(ResolvedSignature s)
+            {
+                if (s.TypeParameters.Any())
+                {
+                    // TODO: throw error
+                    throw new Exception("Signatures cannot contains type parameters");
+                }
+
+                return base.onSignature(s);
+            }
+        }
+
+        internal class MonomorphizationValidationExpression :
+            ExpressionTransformation<Core.ExpressionKindTransformation, MonomorphizationValidationExpressionType>
+        {
+            public MonomorphizationValidationExpression() :
+                base(null, expr => new MonomorphizationValidationExpressionType(expr as MonomorphizationValidationExpression))
```

Comment:
The kind recursion it would prevent is probably not needed for this transform, but this should be filled out anyway as putting null here causes the transformation framework to work in an unintended way. It's my thinking now that this should never be null to avoid strange behavior that is hard to debug. I will fill this argument out.